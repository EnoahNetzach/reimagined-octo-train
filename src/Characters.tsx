import { useCallback, useEffect, useState } from 'react'
import { Link, useParams } from 'react-router-dom'
import { useQuery } from '@apollo/client'
import { gql } from './__generated__'
import { Character } from './__generated__/graphql.ts'
import getCharacterStatusColor from './getCharacterStatusColor.tsx'

const GET_CHARACTERS = gql`
  query GetCharacters($page: Int!, $filter: String) {
    characters(page: $page, filter: { name: $filter }) {
      info {
        count
      }
      results {
        id
        name
        status
        species
        image
        location {
          id
          name
        }
        episode {
          id
          name
        }
      }
    }
  }
`

export default function Characters() {
  const { characterId } = useParams()

  const [filter, setFilter] = useState('')
  const [latestPageLoaded, setLatestPageLoaded] = useState(1)

  useEffect(() => {
    setLatestPageLoaded(1)
  }, [filter])

  const { data, fetchMore, loading } = useQuery(GET_CHARACTERS, { variables: { filter, page: latestPageLoaded } })

  const loadMore = useCallback(async () => {
    try {
      const nextPage = latestPageLoaded + 1
      await fetchMore({ variables: { page: nextPage } })
      setLatestPageLoaded(nextPage)
    } catch {
      // ignore
    }
  }, [fetchMore, latestPageLoaded])

  return (
    <div className="flex gap-4 flex-col">
      <div className="sticky self-start top-0 bg-gray-900 w-full p-2">
        <label>
          Filter: <input onChange={(event) => setFilter(event.target.value)} value={filter} />
        </label>
      </div>

      {loading ? (
        'loading...'
      ) : (
        <>
          {data?.characters?.results
            ?.map((character: Character) =>
              character ? (
                <Link key={character.id} to={`/character/${character.id}`}>
                  <figure
                    aria-selected={characterId === character.id}
                    className={`flex gap-4 flex-row ${characterId === character.id ? 'bg-gray-800' : 'bg-gray-900'}`}
                  >
                    <img alt="" className="object-cover max-h-24 max-w-24" src={character.image} />

                    <figcaption className="flex-grow">
                      <h2 className="text-lg font-bold text-primary">{character.name}</h2>

                      <section>
                        <span
                          aria-description="Status"
                          className={`inline-block rounded-full w-2 h-2 ${getCharacterStatusColor(character)}`}
                        />{' '}
                        {character.status} - {character.species}
                      </section>

                      <section>
                        <h3 className="text-secondary">Last known location:</h3>
                        {character.location?.name ?? 'unknown'}
                      </section>

                      <section>
                        <h3 className="text-secondary">First seen in:</h3>
                        {character.episode[0]?.name}
                      </section>
                    </figcaption>
                  </figure>
                </Link>
              ) : null,
            )
            .filter(Boolean)}

          {data?.characters.info.count > data?.characters?.results.length ? (
            <div>
              <button onClick={loadMore}>Load more</button>
            </div>
          ) : null}
        </>
      )}
    </div>
  )
}
