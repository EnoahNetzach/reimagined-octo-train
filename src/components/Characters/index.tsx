import { useCallback, useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { Link, useParams } from 'react-router-dom'
import { useQuery } from '@apollo/client'
import { gql } from '~/__generated__'
import { Character } from '~/__generated__/graphql'
import getCharacterStatusColor from '~/utils/getCharacterStatusColor'

const GET_CHARACTERS = gql(/* GraphQL */ `
  query GetCharacters($page: Int!, $filter: String) {
    characters(page: $page, filter: { name: $filter }) {
      info {
        count
      }
      results {
        id
        name
        status
        species
        image
        location {
          id
          name
        }
        episode {
          id
          name
        }
      }
    }
  }
`)

export default function Characters() {
  const { t } = useTranslation()
  const { characterId } = useParams()

  const [filter, setFilter] = useState('')
  const [latestPageLoaded, setLatestPageLoaded] = useState(1)
  const [loadingMore, setLoadingMore] = useState(false)

  useEffect(() => {
    setLatestPageLoaded(1)
  }, [filter])

  const { data, fetchMore, loading } = useQuery(GET_CHARACTERS, { variables: { filter, page: latestPageLoaded } })

  const loadMore = useCallback(async () => {
    try {
      const nextPage = latestPageLoaded + 1
      setLoadingMore(true)
      await fetchMore({ variables: { page: nextPage } })
      setLatestPageLoaded(nextPage)
    } catch {
      // ignore
    }
    setLoadingMore(false)
  }, [fetchMore, latestPageLoaded])

  return (
    <div className="flex gap-y-0.5 flex-col">
      <div className="sticky self-start top-0 bg-gray-900 w-full p-2 flex gap-2">
        <label className="self-center" htmlFor="main-search">
          {t('list.filter')}
        </label>
        <input
          className="w-full p-1"
          id="main-search"
          onChange={(event) => setFilter(event.target.value)}
          value={filter}
        />
      </div>

      {loading ? (
        t('loading')
      ) : (
        <nav aria-label="Customer service" className="contents">
          {data?.characters?.results
            ?.filter((character: unknown): character is Character => !!character)
            .map((character: Character) =>
              character ? (
                <Link aria-current={characterId === character.id} key={character.id} to={`/character/${character.id}`}>
                  <figure
                    className={`flex gap-4 flex-row p-2 ${
                      characterId === character.id ? 'bg-gray-800' : 'bg-gray-900'
                    }`}
                    role="row"
                  >
                    <img
                      alt=""
                      className="self-center object-cover h-24 w-24"
                      fetchpriority="low"
                      src={character.image ?? ''}
                    />

                    <figcaption className="flex-grow">
                      <h2 className="heading-title-bold-secondary">{character.name}</h2>

                      <section>
                        <span
                          aria-description={t('character.status')}
                          className={`inline-block rounded-full w-2 h-2 ${getCharacterStatusColor(character)}`}
                        />{' '}
                        {t(`character.status.${character.status?.toLowerCase()}`)} - {character.species}
                      </section>

                      <section>
                        <h3 className="heading-title-tertiary">{t('location.name')}</h3>
                        {character.location?.id ? character.location.name : t('location.unknown')}
                      </section>

                      <section>
                        <h3 className="heading-title-tertiary">{t('episode.first.name')}</h3>
                        {character.episode[0]?.name}
                      </section>
                    </figcaption>
                  </figure>
                </Link>
              ) : null,
            )
            .filter(Boolean)}

          {(data?.characters?.info?.count ?? 0) > (data?.characters?.results?.length ?? 0) ? (
            <div>
              <button disabled={loadingMore} onClick={loadMore}>
                {loadingMore ? t('loading') : t('list.loadMore')}
              </button>
            </div>
          ) : null}
        </nav>
      )}
    </div>
  )
}
