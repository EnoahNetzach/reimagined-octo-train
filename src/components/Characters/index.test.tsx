import * as reactRouterDOM from 'react-router-dom'
import { createRenderer } from 'react-test-renderer/shallow'
import * as apolloClient from '@apollo/client'
import { QueryResult } from '@apollo/client'
import { fireEvent, render } from '@testing-library/react'
import Characters from '.'

const mockCharacter = (n: number) => ({
  episode: [
    {
      air_date: `__character_${n}_episode_air_date__`,
      id: `__character_${n}_episode_id__`,
      name: `__character_${n}_episode_name__`,
    },
  ],
  gender: `__character_${n}_gender__`,
  id: `__character_${n}_id__`,
  image: `__character_${n}_image__`,
  location: {
    dimension: `__character_${n}_location_dimension__`,
    id: `__character_${n}_location_id__`,
    name: `__character_${n}_location_name__`,
    residents: Array.from({ length: 5 }),
    type: `__character_${n}_location_type__`,
  },
  name: `__character_${n}_name__`,
  origin: {
    dimension: `__character_${n}_origin_dimension__`,
    id: `__character_${n}_origin_id__`,
    name: `__character_${n}_origin_name__`,
    residents: Array.from({ length: 6 }),
    type: `__character_${n}_origin_type__`,
  },
  species: `__character_${n}_species__`,
  status: `__character_${n}_status__`,
  type: `__character_${n}_type__`,
})

jest.mock('@apollo/client', () => ({
  useQuery: jest.fn(() => ({
    data: {
      characters: { info: { count: 4 }, results: [mockCharacter(1), mockCharacter(2), mockCharacter(3)] },
    },
    loading: false,
  })),
}))

jest.mock('~/__generated__', () => ({ gql: jest.fn() }))

jest.mock('~/utils/getCharacterStatusColor', () => () => '__character_1_status_color__')

describe('Characters', () => {
  it('mounts', () => {
    const renderer = createRenderer()
    renderer.render(<Characters />)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })

  it('loads', () => {
    jest.spyOn(apolloClient, 'useQuery').mockReturnValueOnce({
      loading: true,
    } as QueryResult<unknown>)

    const renderer = createRenderer()
    renderer.render(<Characters />)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })

  it('has the correct data', () => {
    const { queryByText } = render(<Characters />)
    expect(queryByText('__character_1_name__')).toBeInTheDocument()
    expect(document.querySelector('img[src="__character_1_image__"]')).not.toBeNull()
    expect(queryByText('__character_2_name__')).toBeInTheDocument()
    expect(document.querySelector('img[src="__character_2_image__"]')).not.toBeNull()
    expect(queryByText('__character_3_name__')).toBeInTheDocument()
    expect(document.querySelector('img[src="__character_3_image__"]')).not.toBeNull()
  })

  it('links to the correct profile when clicking to the character', () => {
    const { getByText } = render(<Characters />)

    fireEvent.click(getByText('__character_2_name__'))

    expect(reactRouterDOM.__linkOnClick).toHaveBeenCalledTimes(1)
    expect(reactRouterDOM.__linkOnClick).toHaveBeenCalledWith('/character/__character_2_id__')
  })
})
