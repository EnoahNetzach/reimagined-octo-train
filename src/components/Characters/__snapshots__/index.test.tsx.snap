// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters loads 1`] = `
<div
  className="flex gap-4 flex-col"
>
  <div
    className="sticky self-start top-0 bg-gray-900 w-full p-2"
  >
    <label>
      t(list.filter)
       
      <input
        onChange={[Function]}
        value=""
      />
    </label>
  </div>
  t(loading)
</div>
`;

exports[`Characters mounts 1`] = `
<div
  className="flex gap-4 flex-col"
>
  <div
    className="sticky self-start top-0 bg-gray-900 w-full p-2"
  >
    <label>
      t(list.filter)
       
      <input
        onChange={[Function]}
        value=""
      />
    </label>
  </div>
  <React.Fragment>
    <Link
      to="/character/__character_1_id__"
    >
      <figure
        aria-selected={false}
        className="flex gap-4 flex-row bg-gray-900"
      >
        <img
          alt=""
          className="object-cover max-h-24 max-w-24"
          src="__character_1_image__"
        />
        <figcaption
          className="flex-grow"
        >
          <h2
            className="text-lg font-bold text-primary"
          >
            __character_1_name__
          </h2>
          <section>
            <span
              aria-description="t(character.status)"
              className="inline-block rounded-full w-2 h-2 __character_1_status_color__"
            />
             
            t(character.status.__character_1_status__)
             - 
            __character_1_species__
          </section>
          <section>
            <h3
              className="text-secondary"
            >
              t(location.name)
            </h3>
            __character_1_location_name__
          </section>
          <section>
            <h3
              className="text-secondary"
            >
              t(episode.first.name)
            </h3>
            __character_1_episode_name__
          </section>
        </figcaption>
      </figure>
    </Link>
    <Link
      to="/character/__character_2_id__"
    >
      <figure
        aria-selected={false}
        className="flex gap-4 flex-row bg-gray-900"
      >
        <img
          alt=""
          className="object-cover max-h-24 max-w-24"
          src="__character_2_image__"
        />
        <figcaption
          className="flex-grow"
        >
          <h2
            className="text-lg font-bold text-primary"
          >
            __character_2_name__
          </h2>
          <section>
            <span
              aria-description="t(character.status)"
              className="inline-block rounded-full w-2 h-2 __character_1_status_color__"
            />
             
            t(character.status.__character_2_status__)
             - 
            __character_2_species__
          </section>
          <section>
            <h3
              className="text-secondary"
            >
              t(location.name)
            </h3>
            __character_2_location_name__
          </section>
          <section>
            <h3
              className="text-secondary"
            >
              t(episode.first.name)
            </h3>
            __character_2_episode_name__
          </section>
        </figcaption>
      </figure>
    </Link>
    <Link
      to="/character/__character_3_id__"
    >
      <figure
        aria-selected={false}
        className="flex gap-4 flex-row bg-gray-900"
      >
        <img
          alt=""
          className="object-cover max-h-24 max-w-24"
          src="__character_3_image__"
        />
        <figcaption
          className="flex-grow"
        >
          <h2
            className="text-lg font-bold text-primary"
          >
            __character_3_name__
          </h2>
          <section>
            <span
              aria-description="t(character.status)"
              className="inline-block rounded-full w-2 h-2 __character_1_status_color__"
            />
             
            t(character.status.__character_3_status__)
             - 
            __character_3_species__
          </section>
          <section>
            <h3
              className="text-secondary"
            >
              t(location.name)
            </h3>
            __character_3_location_name__
          </section>
          <section>
            <h3
              className="text-secondary"
            >
              t(episode.first.name)
            </h3>
            __character_3_episode_name__
          </section>
        </figcaption>
      </figure>
    </Link>
    <div>
      <button
        onClick={[Function]}
      >
        t(list.loadMore)
      </button>
    </div>
  </React.Fragment>
</div>
`;
